osu.Game/Screens/Ranking/Statistics/StatisticsPanel.cs

```diff
@@ -74,81 +75,131 @@ private void populateStatistics(ValueChangedEvent<ScoreInfo> score)
+                };
 
-                var localCancellationSource = loadCancellation = new CancellationTokenSource();
-                IBeatmap playableBeatmap = null;
+                bool panelIsEmpty = true;
+                bool panelIsComplete = true;
+                bool hitEventsAvailable = newScore.HitEvents.Count != 0;
 
-                // Todo: The placement of this is temporary. Eventually we'll both generate the playable beatmap _and_ run through it in a background task to generate the hit events.
-                Task.Run(() =>
+                foreach (var row in newScore.Ruleset.CreateInstance().CreateStatisticsForScore(newScore, playableBeatmap))
                 {
-                    playableBeatmap = beatmapManager.GetWorkingBeatmap(newScore.BeatmapInfo).GetPlayableBeatmap(newScore.Ruleset, newScore.Mods);
-                }, loadCancellation.Token).ContinueWith(t => Schedule(() =>
-                {
-                    var rows = new FillFlowContainer
+                    var columnsToDisplay = hitEventsAvailable
+                        ? row.Columns
+                        : row.Columns.Where(c => !c.RequiresHitEvents).ToArray();
+
+                    if (columnsToDisplay.Length < row.Columns.Length)
+                        panelIsComplete = false;
+
+                    if (columnsToDisplay.Any())
+                        panelIsEmpty = false;
+                    else
+                        continue;
+
+                    rows.Add(new GridContainer
                     {
+                        Anchor = Anchor.TopCentre,
+                        Origin = Anchor.TopCentre,
                         RelativeSizeAxes = Axes.X,
                         AutoSizeAxes = Axes.Y,
-                        Anchor = Anchor.Centre,
-                        Origin = Anchor.Centre,
-                        Direction = FillDirection.Vertical,
-                        Spacing = new Vector2(30, 15),
-                        Alpha = 0
-                    };
-
-                    foreach (var row in newScore.Ruleset.CreateInstance().CreateStatisticsForScore(newScore, playableBeatmap))
+                        Content = new[]
+                        {
+                            columnsToDisplay?.Select(c => new StatisticContainer(c)
+                            {
+                                Anchor = Anchor.Centre,
+                                Origin = Anchor.Centre,
+                            }).Cast<Drawable>().ToArray()
+                        },
+                        ColumnDimensions = Enumerable.Range(0, columnsToDisplay.Length)
+                                                     .Select(i => columnsToDisplay[i].Dimension ?? new Dimension()).ToArray(),
+                        RowDimensions = new[] { new Dimension(GridSizeMode.AutoSize) }
+                    });
+                }
+
+                if (!hitEventsAvailable)
+                {
+                    if (panelIsEmpty)
```

Comment:
I would move these right to the top, and have two distinct flows with no `Dispose` calls. It's very weird first making the structure then disposing of it. Let me know if this isn't clear enough and I'll make a diff showing the general expectation.

Reply:
Yeah, I got it. It was just laziness on my part trying to not split the flow.

Reply:
Done. It will first check if all statistic items require hit events. If yes, the placeholder will be created instead of the scroll container.